miniapps/tools/least-squares.cpp

```diff
@@ -0,0 +1,1720 @@
+
+      face_values(p) = shape_self*dofs_at_self;
+      if (has_other) { face_values(p) += shape_other*dofs_at_other; }
+   }
+   if (has_other) { face_values *= 0.5; }
+}
+
+/// @brief Get matrix associated to traces of shape functions on
+/// the face associated to @a face_trans
+void ComputeFaceMatrix(int e_idx,
+                       const FiniteElementSpace& fes,
+                       FaceElementTransformations& face_trans,
+                       const IntegrationRule& ir,
+                       DenseMatrix& e_shape_to_q_face)
+{
+   MFEM_VERIFY(face_trans.Elem1No == e_idx || face_trans.Elem2No == e_idx,
+               "Selected element does not conform face in face_trans!");
+
+   const auto& fe_self = *fes.GetFE(e_idx);
+   const int ndof_self = fe_self.GetDof();
+   auto& e_trans = (face_trans.Elem1No == e_idx)?*face_trans.Elem1:
+                   *face_trans.Elem2;
+
+   Vector shape_self(ndof_self);
+   e_shape_to_q_face.SetSize(ir.GetNPoints(), ndof_self);
+   for (int p = 0; p < ir.GetNPoints(); p++)
+   {
+      const IntegrationPoint& ip = ir.IntPoint(p);
+      face_trans.SetAllIntPoints(&ip);
+      fe_self.CalcPhysShape(e_trans, shape_self);
+      e_shape_to_q_face.SetRow(p, shape_self);
+   }
+}
+
+/// @brief Boilerplate code for getting the orders a priori
+const IntegrationRule& GetCommonIntegrationRule(const FiniteElementSpace&
+                                                fes_src,
+                                                const FiniteElementSpace& fes_dst,
+                                                const FaceElementTransformations& face_trans)
+{
+   const bool has_other = (face_trans.Elem2No >= 0);
+   const FiniteElement* fe_src_self = fes_src.GetFE(face_trans.Elem1No);
+   const FiniteElement* fe_src_other = has_other?fes_src.GetFE(face_trans.Elem2No):
+                                       fe_src_self;
+   const FiniteElement* fe_dst_self = fes_dst.GetFE(face_trans.Elem1No);
+   const FiniteElement* fe_dst_other = has_other?fes_dst.GetFE(face_trans.Elem2No):
+                                       fe_dst_self;
+
+   const int order = face_trans.OrderW() +
+                     std::max(fe_src_self->GetOrder(), fe_dst_self->GetOrder()) +
+                     std::max(fe_src_other->GetOrder(), fe_dst_other->GetOrder());
+   return IntRules.Get(face_trans.GetGeometryType(), order);
+}
+
+///@}
+
+/// @name Reconstruction functions
+///@{
+
+/// @brief Local-averages-based reconstruction
+void AverageReconstruction(Solver& solver,
+                           ParGridFunction& src,
+                           ParGridFunction& dst,
+                           ParMesh& mesh,
+                           IterativeSolverParams& newton,
+                           RegularizationType reg_type = direct,
+                           real_t reg = 0.0,
+                           int print_level = -1,
+                           bool preserve_volumes = false)
```

Comment:
I believe the `ParMesh` object can be obtain through one of the `ParGridFunction` arguments, so I would remove it from the function signature.

Additionally on a first pass read, I believe `print_level` is used more as a boolean flag than an integer print level... if that is the case, I would change the function signature to use a boolean flag.